server/matchmaker/pop_timer.py

```diff
@@ -0,0 +1,68 @@
+import asyncio
+from collections import deque
+from time import time
+from typing import Callable, Deque
+
+import server
+
+from .. import config
+from ..decorators import with_logger
+
+
+@with_logger
+class PopTimer(object):
+    def __init__(self, queue_name: str):
+        self.queue_name = queue_name
+        self.last_queue_amounts: Deque[int] = deque(maxlen=config.QUEUE_POP_TIME_MOVING_AVG_SIZE)
+        self.last_queue_times: Deque[float] = deque(maxlen=config.QUEUE_POP_TIME_MOVING_AVG_SIZE)
+
+        self._last_queue_pop = time()
+        # Optimistically schedule first pop for half of the max pop time
+        self.next_queue_pop = self._last_queue_pop + (config.QUEUE_POP_TIME_MAX / 2)
+
+    async def next_pop(self, get_num_players: Callable[[], int]):
+        """ Wait for the timer to pop. get_num_players needs to return the current
+        number of players in the queue. """
+
+        time_remaining = self.next_queue_pop - time()
+        self._logger.info("Next %s wave happening in %is", self.queue_name, time_remaining)
+        server.stats.timing(
+            "matchmaker.queue.pop", int(time_remaining),
+            tags={'queue_name': self.queue_name}
+        )
+        await asyncio.sleep(time_remaining)
+        num_players = get_num_players()
+        server.stats.gauge(f"matchmaker.queue.{self.queue_name}.players", num_players)
+
+        self._last_queue_pop = time()
+        self.next_queue_pop = self._last_queue_pop + self.time_until_next_pop(
+            num_players, time_remaining
+        )
+
+    def time_until_next_pop(self, num_queued: int, time_queued: float) -> float:
+        """ Calculate how long we should wait for the next queue to pop based
+        on the current rate of ladder queues
+        """
+        self.last_queue_amounts.append(num_queued)
+        self.last_queue_times.append(time_queued)
```

Comment:
do you clear that queues? or they are growing to infinity?

Reply:
python `deque`'s have a builtin feature for max length which I'm setting in `__init__`:

https://docs.python.org/3/library/collections.html#collections.deque

Reply:
And what for is that deque?

If dequeue always has 10 elements, where are you working with that elements?

Reply:
If poptimer is speeding up, by number of people in queue, what happens, when there is not much people on queue? Will poptimer slow down?

Reply:
The deques are used to calculate the moving average of players in the queue. This is basically to smooth out the pop time adjustments since there is a small player base. You can see on the next line it says:
```python
total_players = sum(self.last_queue_amounts)
```

If there are few players in the queue the time will increase up to `config.QUEUE_POP_TIME_MAX`

Reply:
Ok.